### ISSUE TITLE
Deleted pod not removed from EDS

### ISSUE THREAD

**Issue by Stono** [istio/istio#10580]
Potentially a dupe of https://github.com/istio/istio/issues/9254 and potentially a long standing issue that I've not noticed before.  We've recently added alerts which monitor EDS, however we restart pilot nightly to "rebalance" it, so that may explain why we haven't noticed this before (nobody is eyeballing EDS and we restart pilot daily).

<img width="637" alt="screenshot 2018-12-19 at 17 42 53" src="https://user-images.githubusercontent.com/1486729/50237786-a5f87c00-03b5-11e9-82be-133ba979eae1.png">

I'm running `1.0.5`.

**Summary**
Pilot EDS instances reports pods that no longer exist, for example:

```
istioctl proxy-status | grep sourcing
sourcing-client-5fc9c9dc76-wzgls.sourcing-client                                                 SYNCED     SYNCED     STALE (100%)      SYNCED     istio-pilot-6449ddbbcc-2z9nh     1.0.2
sourcing-client-8664c9bdbc-7h49x.sourcing-client                                                 SYNCED     SYNCED     SYNCED (100%)     SYNCED     istio-pilot-6449ddbbcc-rwnfk     1.0.2
sourcing-client-8664c9bdbc-f9mbc.sourcing-client                                                 SYNCED     SYNCED     SYNCED (100%)     SYNCED     istio-pilot-6449ddbbcc-rwnfk     1.0.2
sourcing-client-8664c9bdbc-n8jtx.sourcing-client                                                 SYNCED     SYNCED     SYNCED (100%)     SYNCED     istio-pilot-6449ddbbcc-2z9nh     1.0.2
```

Note that `sourcing-client-5fc9c9dc76-wzgls.sourcing-client`, marked as stale, doesn't exist and hasn't existed for about 10 minutes:

```
❯ kubectl -n sourcing-client get pods
NAME                               READY   STATUS    RESTARTS   AGE
sourcing-client-8664c9bdbc-7h49x   2/2     Running   0          7m
sourcing-client-8664c9bdbc-f9mbc   2/2     Running   0          7m
sourcing-client-8664c9bdbc-n8jtx   2/2     Running   0          7m
```

Restarting pilot fixed the problem and removed the stale data.

I've provided full debug logs from all pilots, as well as cluster information from before and after the restart to @costinm and @mandarjog via slack as they contain sensitive information.

**Comment by Stono** [comment]
Did a bit of digging to see what, if any impact this caused.  Unfortunately the app is *very* low traffic and only received around 300 requests during the 20 minute window in which pilot had this state.

If i query the source reporter (`istio_requests_total{reporter="source", destination_app ="sourcing-client", response_code=~"5.*"}`) there doesn't appear to be any `5xx` requests.

Interestingly if i query the `reporter="destination"` instead, I see just a single 503.

The source app in this instance is nginx, looking at the envoy logs on the two instances of nginx we're running, I see no retries in envoy, which I would expect on gateway errors due to our retry configuration.

```
www-data@ingress-nginx-internal-controller-5d7c98684d-gj89t:/etc/nginx$ curl -s http://127.0.0.1:15000/stats | grep sourcing-client | grep retry
cluster.outbound|80||app.sourcing-client.svc.cluster.local.retry_or_shadow_abandoned: 0
cluster.outbound|80||app.sourcing-client.svc.cluster.local.upstream_rq_retry: 0
cluster.outbound|80||app.sourcing-client.svc.cluster.local.upstream_rq_retry_overflow: 0
cluster.outbound|80||app.sourcing-client.svc.cluster.local.upstream_rq_retry_success: 0
```

and:
```
www-data@ingress-nginx-internal-controller-5d7c98684d-gj89t:/etc/nginx$ curl -s http://127.0.0.1:15000/stats | grep sourcing-client | grep retry
cluster.outbound|80||app.sourcing-client.svc.cluster.local.retry_or_shadow_abandoned: 0
cluster.outbound|80||app.sourcing-client.svc.cluster.local.upstream_rq_retry: 0
cluster.outbound|80||app.sourcing-client.svc.cluster.local.upstream_rq_retry_overflow: 0
cluster.outbound|80||app.sourcing-client.svc.cluster.local.upstream_rq_retry_success: 0
```

and nginx `nginx_ingress_controller_requests{namespace="sourcing-client", status=~"5.*"}` reports no errors.

**Comment by Stono** [comment]
Ahh; the uptime of nginx in this case shows that it was restarted *after* this incident therefore the envoy retry logs above are moot (and unfortunately we currently don't scrape them to prometheus)

**Comment by hzxuzhonghu** [comment]
@Stono Is this fixed?

**Comment by Stono** [comment]
Yeah it was fixed by a patch in 1.0.6, we're testing 1.1.0 at the moment so will explicitly look for this

**Comment by hzxuzhonghu** [comment]
OK, close when you verified
